File: 173. Binary Search Tree Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
private:
    vector <int> nums;
    stack <TreeNode*> s;
public:
    BSTIterator(TreeNode *root) {
        if (root == NULL) return;
        s.push(root);
        while (!s.empty())
        {
            while (root != NULL)
            {
                s.push(root);
                root = root->left;
            }
            root = s.top(); s.pop();
            nums.push_back(root->val);
            root = root->right;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return nums.size() > 1 ? true : false;
    }

    /** @return the next smallest number */
    int next() {
        int temp = nums[0];
        nums.erase(nums.begin());
        return temp;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: 173. Binary Search Tree Iterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.nums = []
        self.stack = [root]
        if not root:
            return
        while self.stack:
            while root:
                self.stack += [root]
                root = root.left
            root = self.stack.pop()
            self.nums += [root.val]
            root = root.right

    def hasNext(self):
        """
        :rtype: bool
        """
        return True if len(self.nums) > 1 else False

    def next(self):
        """
        :rtype: int
        """
        return self.nums.pop(0)

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())
